src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -4040,26 +4259,321 @@ ColocationGroupCreateCommandList(void)
 							 "NULL, NULL)");
 		}
 
-		colocationTuple = systable_getnext_ordered(scanDescriptor, ForwardScanDirection);
+		appendStringInfo(colocationGroupCreateCommand,
+						 ") SELECT pg_catalog.citus_internal_add_colocation_metadata("
+						 "colocationid, shardcount, replicationfactor, "
+						 "distributioncolumntype, coalesce(c.oid, 0)) "
+						 "FROM colocation_group_data d LEFT JOIN pg_collation c "
+						 "ON (d.distributioncolumncollationname = c.collname "
+						 "AND d.distributioncolumncollationschema::regnamespace"
+						 " = c.collnamespace)");
+
+		SendCommandListToActivatedNodes(context, list_make1(
+											colocationGroupCreateCommand->data));
 	}
+	MemoryContextSwitchTo(oldContext);
 
 	systable_endscan_ordered(scanDescriptor);
 	index_close(colocationIdIndexRel, AccessShareLock);
 	table_close(pgDistColocation, AccessShareLock);
+}
+
 
-	if (!hasColocations)
+/*
+ * SendDependencyCreationCommands sends dependency creation commands to workers
+ * with transactional or nontransactional mode according to transactionMode
+ * inside metadataSyncContext.
+ */
+void
+SendDependencyCreationCommands(MetadataSyncContext *context)
+{
+	MemoryContext oldContext = MemoryContextSwitchTo(context->context);
+
+	/*
+	 * collect all dependencies in creation order and get their ddl commands
+	 */
+	List *dependencies = GetDistributedObjectAddressList();
+
+	/*
+	 * Depending on changes in the environment, such as the enable_metadata_sync guc
+	 * there might be objects in the distributed object address list that should currently
+	 * not be propagated by citus as they are 'not supported'.
+	 */
+	dependencies = FilterObjectAddressListByPredicate(dependencies,
+													  &SupportedDependencyByCitus);
+
+	dependencies = OrderObjectAddressListInDependencyOrder(dependencies);
+
+	MemoryContext commandsContext = AllocSetContextCreate(context->context,
```

Comment:
maybe `commandsContext` -> `perDependencyContext` or such.


Also, here is the only place where we have an additional context. Could use a comment on why we have this when we do `MemoryContextSwitchTo(context->context)`

Reply:
we get rid of context switch after refactor. Resetting memory causes SEGFAULT only for index scan. But we do not need index scan as we already scan the whole table. After replacing index scan with heap scan, the problem is gone. We donot need another subcontext.